utils/portalloc/src/data/dto.rs

```diff
@@ -32,14 +36,63 @@ impl Default for RootData {
 }
 
 impl RootData {
+    pub fn get_free_port_range(&mut self, range_size: u16) -> u16 {
+        let mut base_port: u16 = self.next;
+        'retry: loop {
+            trace!(target: LOG_PORT_ALLOC, base_port, range_size, "Checking a port");
+            if HIGH < base_port {
+                self.reclaim(now_ts());
+                base_port = LOW;
+            }
+            let range = base_port..base_port + range_size;
+            if let Some(next_port) = self.contains(range.clone()) {
+                warn!(
+                    base_port,
+                    range_size,
+                    "Could not use a port (already reserved). Will try a different range."
+                );
+                base_port = next_port;
+                continue 'retry;
+            }
+
+            for port in range.clone() {
+                match TcpListener::bind(("127.0.0.1", port)) {
+                    Err(error) => {
+                        warn!(
+                            ?error,
+                            port, "Could not use a port. Will try a different range"
+                        );
+                        base_port = port + 1;
+                        continue 'retry;
+                    }
+                    Ok(l) => l,
+                };
+            }
+
+            const ALLOCATION_TIME_SECS: u64 = 120;
```

Comment:
If you're refactoring it (which is a good idea), I would move the stuff from this line downwards to a separate function, so the logic here can be one larger loop-y thing on immutable `&self` and then mutations live outside. Unless I'm missing something here that makes it necessary to remain mutable.

Reply:
I moved this code into `RootData::insert`. But the loop here calls `RootData::reclaim` which requires `&mut self`.